hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java

```diff
@@ -54,192 +61,217 @@
 
-    public void preHandle(@NonNull final Iterator<Transaction> itr, @NonNull final HederaState state) {
-        // Each transaction in the event will go through pre-handle using a background thread
-        // from the executor service. The Future representing that work is stored on the
-        // platform transaction. The HandleTransactionWorkflow will pull this future back
-        // out and use it to block until the pre handle work is done, if needed.
-        final ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
-        while (itr.hasNext()) {
-            final var platformTx = itr.next();
-            final var future = runner.apply(() -> {
-                try {
-                    final var metadata = securePreHandle(state, platformTx);
-                    platformTx.setMetadata(metadata);
-                } catch (final PreCheckException e) {
-                    platformTx.setMetadata(createInvalidResult(e.responseCode()));
-                }
-            });
-            futures.add(future);
+        requireNonNull(readableStoreFactory);
+        requireNonNull(creator);
+        requireNonNull(transactions);
+
+        // Used for looking up payer account information.
+        final var accountStore = readableStoreFactory.createStore(ReadableAccountStore.class);
+
+        // Using the executor service, submit a task for each transaction in the event.
+        final var tasks = new ArrayList<WorkItem>(1000); // Some arbitrary number
+        while (transactions.hasNext()) {
+            // Skip platform (system) transactions.
+            final var platformTx = transactions.next();
+            if (platformTx.isSystem()) continue;
+
+            // Submit the task to the executor service and put the resulting Future as the metadata on the transaction
+            final var future =
+                    exe.submit(() -> preHandleTransaction(creator, readableStoreFactory, accountStore, platformTx));
+            tasks.add(new WorkItem(platformTx, future));
         }
 
-        // wait until all transactions were processed before returning
-        final CompletableFuture<?>[] array = futures.toArray(new CompletableFuture<?>[0]);
-        CompletableFuture.allOf(array).join();
+        // Waits for all the background threads to complete their work and stores the resulting PreHandleResult
+        // as the transaction's metadata.
+        for (final var task : tasks) {
+            try {
+                // FUTURE: Enhance this so the timeout duration is a configuration property (and then update the
+                // unit tests, so they use a much shorter configuration, so test execution is faster).
+                final var config = configProvider.getConfiguration().getConfigData(PreHandleConfig.class);
+                final var timeout = config.timeout();
+                final var result = task.future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
This could potentially result in waiting many times the timeout if several items are delayed close to, but not over, the timeout limit.
Rather than calling get with a timeout in a loop, we might be better off joining all of the threads with a timeout, then get all the results knowing they're all immediately available.
Changes needed:
 * change the Futures to be CompletableFutures
 * Use `CompletableFuture.allOf(arrayOfFutures).get()` (returns void) to wait for all to complete before entering the loop on line 150; when that call returns all of the futures are guaranteed to have results available (or throw ExecutionException).
 * Adjust to ensure the methods that return Futures do not return null (null causes issues with allOf).
Note, you could also use `CompletableFuture.allOf(arrayOfFutures).thenApplyAsync(this::methodHandleToHandleEachFuture)` to run the loop for items as they arrive (rather than once all are ready).

Reply:
The problem is the future is not mine, it is created by the `Executor`, so I cannot complete it. I wish we had `allOf` for regular futures :-(.

Reply:
You can pass the executor to the CompletableFuture call, which avoids the Executor creating a regular Future.  See the example code I wrote above.

Reply:
Oh, this is very interesting. I did not know this. I'm not sure how I feel about it yet. I have two reactions:

1. We should use `CompletableFuture` everywhere we use Future because `CompletableFuture` has all this awesome API (`join`, `runAfterBoth`, ... and a million other things) by virtue of `CompletionStage`.
2. `CompletableFuture` is the wrong abstraction. The Java team should have made `Future` extend `CompletionStage`

The problem with `CompletableFuture` is it *does* expose API for completing the future. It seems like a bad idea for object 1 to create a future, give it to object 2, and allow object 2 to complete the future, irrespective of whatever background work was actually going on, as the following test demonstrates:

```
    @Test
    void completableFutureOverride() throws ExecutionException, InterruptedException {
        final var exe = Executors.newSingleThreadExecutor();
        final var cf = CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            return 1; // The actual value computed by the background thread
        }, exe);

        System.out.println(cf.complete(2)); // prints "true", and supplies a different value than the computing thread
        System.out.println(cf.get()); // prints '2'. It overrode whatever the background thread was doing.
    }
```

That seems wrong and could be dangerous in a program. Sometimes this might be useful and desirable behavior, when you really *want* the ability to override whatever computation was happening in the background. It seems like the only reason for using `CompletableFuture` is because `Future` doesn't extend from `CompletionStage`. Maybe we should have our own `Future` subclass that does, and use that everywhere (it could even be implemented with `CompletableFuture` behind the scenes but hide the dangerous parts).